File: cpp/src/system/variables/eos.cpp
#include "../../etc/globals.h"
#include "../functions/matrices.h"
#include "../functions/vectors.h"
#include "../objects/gpr_objects.h"
#include "mg.h"
#include "shear.h"
#include <cmath>

double E_1(double ρ, double p, Par &MP) {
  // Returns the microscale energy under the MG EOS
  double Γ = Γ_MG(ρ, MP);
  double pr = p_ref(ρ, MP);
  double er = e_ref(ρ, MP);
  return er + (p - pr) / (ρ * Γ);
}

double E_2A(double ρ, Mat3_3r A, Par &MP) {
  // Returns the mesoscale energy dependent on the distortion
  double C0 = C_0(ρ, MP);
  return C0 / 4 * devGsq(A);
}

double E_2J(Vec3r J, Par &MP) {
  // Returns the mesoscale energy dependent on the thermal impulse
  return MP.cα2 * J.squaredNorm() / 2;
}

double E_3(Vec3r v) {
  // Returns the macroscale kinetic energy
  return v.squaredNorm() / 2;
}

double E_R(double λ, Par &MP) {
  // Returns the microscale energy corresponding to the chemical energy in a
  // reactive material
  return MP.Qc * (λ - 1);
}

double total_energy(double ρ, double p, Vec3r v, Par &MP) {
  double E = E_1(ρ, p, MP) + E_3(v);
  return E;
}

double total_energy(double ρ, double p, Mat3_3r A, Vec3r v, Par &MP) {
  double E = E_1(ρ, p, MP) + E_3(v);
  E += E_2A(ρ, A, MP);
  return E;
}

double total_energy(double ρ, double p, Mat3_3r A, Vec3r J, Vec3r v, Par &MP) {
  double E = E_1(ρ, p, MP) + E_3(v);
  E += E_2A(ρ, A, MP);
  E += E_2J(J, MP);
  return E;
}

double total_energy(double ρ, double p, Mat3_3r A, Vec3r J, Vec3r v, double λ,
                    Par &MP) {
  double E = E_1(ρ, p, MP) + E_3(v);
  E += E_2A(ρ, A, MP);
  E += E_2J(J, MP);
  E += E_R(λ, MP);
  return E;
}

File: python/gpr/vars/eos.py
from gpr.misc.functions import dev, gram, L2_1D, L2_2D
from gpr.vars import mg
from gpr.vars.shear import C_0


def E_1(ρ, p, MP):
    """ Returns the microscale energy
    """
    return mg.internal_energy(ρ, p, MP)


def E_2A(ρ, A, MP):
    """ Returns the mesoscale energy dependent on the distortion
    """
    C0 = C_0(ρ, MP)
    G = gram(A)
    return C0 / 4 * L2_2D(dev(G))


def E_2J(J, MP):
    """ Returns the mesoscale energy dependent on the thermal impulse
    """
    cα2 = MP.cα2
    return cα2 / 2 * L2_1D(J)


def E_3(v):
    """ Returns the macroscale kinetic energy
    """
    return L2_1D(v) / 2


def E_R(λ, MP):
    """ Returns the microscale energy corresponding to the chemical energy in a
        reactive material
    """
    Qc = MP.Qc
    return Qc * (λ - 1)


def total_energy(ρ, p, v, A, J, λ, MP):
    """ Returns the total energy
    """
    ret = E_1(ρ, p, MP)

    if A is not None:
        ret += E_2A(ρ, A, MP)

    if J is not None:
        ret += E_2J(J, MP)

    if λ is not None:
        ret += E_R(λ, MP)

    ret += E_3(v)

    return ret
